types/src/transaction/automated_transaction.rs

```diff
@@ -131,10 +138,213 @@ impl AutomatedTransaction {
             )
         })
     }
+
+    /// Returns transaction TTL since base_timestamp if the transaction expiry time is in the future,
+    /// otherwise None.
+    pub fn duration_since(&self, base_timestamp: u64) -> Option<u64> {
+        self.expiration_timestamp_secs().checked_sub(base_timestamp)
+
+    }
 }
 
 impl From<AutomatedTransaction> for Transaction {
     fn from(value: AutomatedTransaction) -> Self {
         Transaction::AutomatedTransaction(value)
     }
 }
+
+macro_rules! value_or_missing {
+    ($value: ident , $message: literal) => {
+        match $value {
+            Some(v) => v,
+            None => return BuilderResult::missing_value($message),
+        }
+    };
+}
+#[derive(Clone, Debug)]
+pub enum BuilderResult {
+    Success(AutomatedTransaction),
+    GasPriceThresholdExceeded { threshold: u64, value: u64 },
+    MissingValue(&'static str),
+}
+
+impl BuilderResult {
+    pub fn success(txn: AutomatedTransaction) -> BuilderResult {
+        Self::Success(txn)
+    }
+
+    pub fn gas_price_threshold_exceeded(threshold: u64, value: u64) -> BuilderResult {
+        Self::GasPriceThresholdExceeded { threshold, value }
+    }
+    pub fn missing_value(missing: &'static str) -> BuilderResult {
+        Self::MissingValue(missing)
+    }
+}
+
+/// Builder interface for [AutomatedTransaction]
+#[derive(Clone, Debug, Default)]
+pub struct AutomatedTransactionBuilder {
+    /// Gas unit price threshold. Default to 0.
```

Comment:
Should the default be same as chain min?

Reply:
Unfortunantly `GAS_UNIT_PRICE` constant is not accessible in `types` crate.

Reply:
@aregng , ok for now we can leave it. We should develop `chain_info.move` module that exposes chain parameters to Move layers (read-only). But that could be a separate issue/project.